### PR TITLE
Allow a precision to be specified for the 'G' format in Utf8Formatter

### PR DESCRIPTION
**Pull request by tannergooding**
This resolves https://github.com/dotnet/corefx/issues/38416

#### Description

Removes an artificial limitation in the Utf8Formatter that was preventing a precision from being given for the 'G' format specifier.

#### Customer Impact

Taking this fix will allow users utilizing System.Text.Json on .NET Framework or .NET Core versions prior to 3.0 to serialize single-precision or double-precision floating-point values in a manner which allows them to be roundtripped to the original value when deserializing.

#### Regression

No

#### Packaging Reviewed

Reviewed by @ericstj 

#### Risk

Minimal. The Utf8Formatter for `float`/`double` is merely forwarding to the `float.ToString` and `double.ToString` implementation. The same change (but allowing any format specifier) was already taken for .NET Core 3.0.

### CODE REVIEW

**Comment by danmosemsft** [comment]
@ericstj can you assess the risk of re-shipping this package? I see we shipped 4.5.3 recently, so is this a well trodden path already?
https://www.nuget.org/packages/System.Memory/

**Comment by danmosemsft** [comment]
Also tests please :)

**Comment by ericstj** [comment]
Low risk to reship so long as you follow the template: https://github.com/dotnet/corefx/commit/20e9f7ae1235f11bcd61e3e6826e5256b6d8cab6

File: src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Float.cs [hunk 1]
```diff
@@ -76,10 +76,6 @@ public static bool TryFormat(float value, Span<byte> destination, out int bytesW
             {
                 case 'g':
                 case 'G':
-                    if (format.Precision != StandardFormat.NoPrecision)
-                        throw new NotSupportedException(SR.Argument_GWithPrecisionNotSupported);
-                    break;
```
Comment (stephentoub) [inline src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Float.cs]: There's a similar check for decimal:
https://github.com/dotnet/corefx/blob/b86783ef9525f22b0576278939264897464f9bbd/src/Common/src/CoreLib/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Decimal.cs#L44
Is that one ok, or should it be removed as well?

**Comment by stephentoub** [comment]
Tests?

**Comment by tannergooding** [comment]
> Tests?

Will get them added tomorrow. There were a number of other things that had to get done today.

File: src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Float.cs (hunk 1 above)
Comment (tannergooding) [reply src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Float.cs]: That check still exists in .NET Core 3.0 as well.

I don't believe it is required for `decimal` however, as I believe it produces a round-trippable result by default.

**Comment by tannergooding** [comment]
@stephentoub, @safern. Are tests not running for this branch?

I would have expected almost all of these legs to have failed the `System.Memory` tests....

**Comment by stephentoub** [comment]
> Are tests not running for this branch?

They don't appear to be, e.g. there's no Send to Helix step in https://dev.azure.com/dnceng/public/_build/results?buildId=219112&view=logs.  I don't know why.  @wtgodbe?

**Comment by safern** [comment]
>They don't appear to be, e.g. there's no Send to Helix step in https://dev.azure.com/dnceng/public/_build/results?buildId=219112&view=logs. I don't know why. @wtgodbe?

Tests in release/2.1 and release/2.2 branches are ran on official builds only. That is a TODO that is missing on release branches that @wtgodbe or @mmitche would know when/if is going to be done.

https://github.com/dotnet/corefx/blob/release/2.1/.azure-ci.yml#L91

File: src/System.Memory/tests/ParsersAndFormatters/SupportedFormats.cs [hunk 2]
```diff
@@ -69,8 +69,8 @@ public static IEnumerable<SupportedFormat> FloatingPointFormats
         {
             get
             {
-                yield return new SupportedFormat('G', supportsPrecision: false) { IsDefault = true };
-                yield return new SupportedFormat('g', supportsPrecision: false) { FormatSynonymFor = 'G', ParseSynonymFor = 'G' };
+                yield return new SupportedFormat('G', supportsPrecision: true) { IsDefault = true };
```
Comment (tannergooding) [inline src/System.Memory/tests/ParsersAndFormatters/SupportedFormats.cs]: This just enables the existing logic to do the precision checks.

In the 3.0 branch, this was all replaced with more explicit tests that are shared between x86 and x64 instead.

**Comment by tannergooding** [comment]
> Tests in release/2.1 and release/2.2 branches are ran on official builds only

@safern, is there a way to manually trigger them? I'd rather not have to download/install VS2017 to build/test locally if I don't need to.

**Comment by safern** [comment]
> @safern, is there a way to manually trigger them? I'd rather not have to download/install VS2017 to build/test locally if I don't need to.

You could build them locally and send a private run to helix, but for that you would need to build locally. 

I can't think of an easy way to trigger them from these branches. Maybe @mmitche knows.

**Comment by wtgodbe** [comment]
Your best bet might be to run a pipebuild against this branch of CoreFx, with publishing disabled. If I'm remembering correctly, there's a property you can set in the pipebuild (something like "PrivateBuild") which will do that. Looking now

**Comment by wtgodbe** [comment]
OK, I think what you need to do is push your changes to a new branch of https://devdiv.visualstudio.com/DevDiv/_git/DotNet-CoreFX-Trusted, then run a build of https://devdiv.visualstudio.com/DevDiv/_build?definitionId=4618, setting `PB_PrivateRun` to `true`, and pointing the build to your new branch. It may be best to wait for confirmation from @mmitche that nothing else is needed to ensure the world doesn't explode, but I'm about 90% sure that this should work 😆

**Comment by danmosemsft** [comment]
> Taking this fix will allow users utilizing System.Text.Json on .NET Framework or .NET Core versions prior to 3.0 to serialize single-precision or double-precision floating-point values in a manner which allows them to be roundtripped to the original value when deserializing.

For shiproom, it would be helpful if you can share any customer requests or feedback you have asking for this (presumably from customers that have to remain on LTS). Otherwise we could postpone it until we get such feedback, or put it in 2.2 and hold on 2.1.

**Comment by tannergooding** [comment]
I don't think putting it into 2.2 works. This is an OOB package and having full framework and 2.2 have one behavior but 2.1 have a difference one would be weird/confusing.

> customer requests or feedback you have asking for this (presumably from customers that have to remain on LTS).

I don't think we have had explicit feedback yet. This is going to be an issue though as it means that you cannot serialize a `double`/`float` and retrieve back the original value on older frameworks (the default on older frameworks is effectively `G15`, which means you truncate bits and can have results off by a large order of magnitude).

For example, `Math.PI` and the immediate next/previous representable values all serialize to the same string by default on the older frameworks. As do, `2147483647`, `2147483646`, and `2147483648`, etc.

**Comment by mmitche** [comment]
Yeah, pushing to the internal repo and building that using the pipebuild def is probably your best bet.

**Comment by tannergooding** [comment]
@ericstj, could you signoff on the packaging commit I just did so I can send the mail to shiproom?

File: src/System.Memory/dir.props [hunk 3]
```diff
@@ -2,7 +2,7 @@
 <Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <Import Project="..\dir.props" />
   <PropertyGroup>
-    <AssemblyVersion>4.0.1.1</AssemblyVersion>
+    <AssemblyVersion>4.0.1.2</AssemblyVersion>
     <PackageVersion>4.5.1</PackageVersion>
```
Comment (ericstj) [inline src/System.Memory/dir.props]: nit: not from your commit, but we should either use this value or delete it.

Comment (tannergooding) [reply src/System.Memory/dir.props]: Do you feel it should be included as part of this PR or that it should be left "as-is" in the servicing branch here and removed in master?

**Comment by wtgodbe** [comment]
@tannergooding it may be too late to get this into the July 2.x release, the builds began yesterday - CC @vivmishra @mmitche

**Comment by danmosemsft** [comment]
I am going to close this and see what customer feedback we get.

**Pull request closed without merging**